modules/images/webp-uploads/load.php

```diff
@@ -5,40 +5,294 @@
+	}
+
+	// All subsizes are created out of the attached file.
+	$file = get_attached_file( $attachment_id );
+
+	// File does not exist.
+	if ( ! file_exists( $file ) ) {
+		return new WP_Error( 'image_file_size_not_found', __( 'The provided size does not have a valid image file.', 'performance-lab' ) );
+	}
+
+	// Create the subsizes out of the attached file.
+	$editor = wp_get_image_editor( $file );
+
+	if ( is_wp_error( $editor ) ) {
+		return $editor;
+	}
+
+	$allowed_mimes = array_flip( wp_get_mime_types() );
+	if ( ! array_key_exists( $mime, $allowed_mimes ) || ! is_string( $allowed_mimes[ $mime ] ) ) {
+		return new WP_Error( 'image_mime_type_invalid', __( 'The provided mime type is not allowed.', 'performance-lab' ) );
+	}
+
+	if ( ! wp_image_editor_supports( array( 'mime_type' => $mime ) ) ) {
+		return new WP_Error( 'image_mime_type_not_supported', __( 'The provided mime type is not supported.', 'performance-lab' ) );
+	}
+
+	$extension = explode( '|', $allowed_mimes[ $mime ] );
+	$extension = reset( $extension );
+
+	$width  = null;
+	$height = null;
+	$crop   = false;
+
+	if ( array_key_exists( 'width', $metadata['sizes'][ $size ] ) ) {
+		$width = $metadata['sizes'][ $size ]['width'];
+	} elseif ( array_key_exists( 'width', $sizes[ $size ] ) ) {
+		$width = $sizes[ $size ];
+	}
+
+	if ( array_key_exists( 'height', $metadata['sizes'][ $size ] ) ) {
+		$height = $metadata['sizes'][ $size ]['height'];
+	} elseif ( array_key_exists( 'width', $sizes[ $size ] ) ) {
+		$height = $sizes[ $size ];
+	}
+
+	if ( array_key_exists( 'crop', $sizes[ $size ] ) ) {
+		$crop = (bool) $sizes[ $size ]['crop'];
+	}
+
+	$editor->resize( $width, $height, $crop );
+	$filename = $editor->generate_filename( null, null, $extension );
+	$filename = preg_replace( '/-(scaled|rotated|imagifyresized)/', '', $filename );
+	$image    = $editor->save( $filename, $mime );
+
+	if ( is_wp_error( $image ) ) {
+		return $image;
 	}
 
-	// WebP lossless support is still limited on servers, so only apply to JPEGs.
-	if ( 'image/jpeg' !== $mime_type ) {
-		return $output_format;
+	if ( empty( $image['file'] ) ) {
+		return new WP_Error( 'image_file_not_present', __( 'The file key is not present on the image data', 'performance-lab' ) );
 	}
 
-	// Skip conversion when creating the `-scaled` image (for large image uploads).
-	if ( preg_match( '/-scaled\..{3}.?$/', $filename ) ) {
-		return $output_format;
+	if ( ! isset( $metadata['sizes'][ $size ]['sources'] ) || ! is_array( $metadata['sizes'][ $size ]['sources'] ) ) {
+		$metadata['sizes'][ $size ]['sources'] = array();
 	}
 
-	$output_format['image/jpeg'] = 'image/webp';
+	$metadata['sizes'][ $size ]['sources'][ $mime ] = array(
+		'file'     => $image['file'],
+		'filesize' => array_key_exists( 'path', $image ) ? filesize( $image['path'] ) : 0,
+	);
+
+	return wp_update_attachment_metadata( $attachment_id, $metadata );
```

Comment:
See my comment above, we should return the new source here so that it can also be processed outside, so that the data in the above function remains intact.
```suggestion
	$new_source = array(
		'file'     => $image['file'],
		'filesize' => isset( $image['path'] ) ? filesize( $image['path'] ) : 0,
	);

	// Update metadata right away with the new source in case the process fails later.
	$metadata['sizes'][ $size ]['sources'][ $mime ] = $new_source;
	wp_update_attachment_metadata( $attachment_id, $metadata );

	return $new_source;
```

Reply:
Right, this behavior is to allow an `async` mechanism rather than a single request.

Reply:
See my comment above - I think we should change that to use a single request while supporting retry attempts, thus following how WP core does it.